libnss_determined/src/group.c

```diff
@@ -0,0 +1,99 @@
+#include "libnss_determined.h"
+
+/*
+group database functions for listing groups:
+
+    These are called to list groups in the database:
+        _nss_determined_setgrent
+        _nss_determined_getgrent_r
+        _nss_determined_endgrent
+
+    This is called to look up a group by name:
+        _nss_determined_getgrnam_r
+
+    This is called to look up a group by uid:
+        _nss_determined_getgrgid_r
+*/
+
+FILE *g_group_f = NULL;
+
+enum nss_status _nss_determined_setgrent (int stayopen) {
+    DEBUG("called, stayopen=%d\n", stayopen);
+
+    det_nss_status_t det_nss_status = det_fopen(GROUPFILE, &g_group_f);
+
+    return set_errnop_and_return_nss_status(det_nss_status, &errno);
+}
+
+enum nss_status _nss_determined_getgrent_r (struct group *result, char *buf,
+        size_t buflen, int *errnop) {
+    DEBUG("called\n");
+
+    det_nss_status_t det_nss_status = getent_any(
+        GROUPFILE, &g_group_f, parse_group_line, (void*)result, buf, buflen
+    );
+
+    if(det_nss_status){
+        *result = (struct group){0};
+        memset(buf, 0, buflen);
+    }
+
+    return set_errnop_and_return_nss_status(det_nss_status, errnop);
+}
+
+enum nss_status _nss_determined_endgrent (void){
+    DEBUG("called\n");
+    if(g_group_f) fclose(g_group_f);
+    return NSS_STATUS_SUCCESS;
+}
+
+
+bool group_match_name(void *id, void *result_any){
+    char *name = id;
+    struct group *result = result_any;
+    return strcmp(name, result->gr_name) == 0;
+}
+
+// Prototype inferred from gilbc/nis/nss_compat/compat-grp.c.
+enum nss_status _nss_determined_getgrnam_r(const char *name,
+        struct group *result, char *buf, size_t buflen, int *errnop){
+    DEBUG("called, name=%s\n", name);
+
+    det_nss_status_t det_nss_status = searchent_any(
+        GROUPFILE, parse_group_line, group_match_name, (void*)name,
+        (void*)result, buf, buflen
+    );
+
+    if(det_nss_status){
+        *result = (struct group){0};
```

Comment:
non-blocking: If you're willing to either (1) add a `result_size` argument to `searchent_any` and use `memset` for zeroing the output or (2) add a `zero_result` argument and a new set of corresponding functions, I think you can turn all the functions in this family into a single line (well, except for the `DEBUG` call).